File: sdk/typescript/examples/basic-chat.ts
/**
 * Minimal end-to-end example: spawn harnessd, create a session, send one
 * prompt, and print streamed text deltas until the run completes.
 *
 * Build the workspace first (`cargo build --release`), then run:
 *
 *   ANTHROPIC_API_KEY=sk-... node --experimental-strip-types examples/basic-chat.ts
 *
 * or compile with `npm run build` and run the emitted JS.
 */

import { HarnessClient } from "../src/client.js";
import { StdioSidecarTransport } from "../src/transport.js";

async function main(): Promise<void> {
  const transport = new StdioSidecarTransport({
    command: "../../../target/release/harnessd",
    onStderrLine: (line) => console.error(`[harnessd] ${line}`),
  });

  const client = await HarnessClient.connect(transport);

  const session = await client.createSession({
    workspaceRoot: process.cwd(),
    integration: "anthropic",
    integrationConfig: {},
  });

  await session.subscribe((event) => {
    const payload = event.event;
    if (typeof payload !== "string" && "AssistantTextDelta" in payload) {
      process.stdout.write(payload.AssistantTextDelta.delta);
    } else if (typeof payload !== "string" && "Completed" in payload) {
      process.stdout.write("\n");
    } else if (typeof payload !== "string" && "Failed" in payload) {
      console.error("run failed:", payload.Failed.error);
    }
  });

  await session.prompt("In one short sentence, what is a Rust trait?");

  await new Promise((resolve) => setTimeout(resolve, 15_000));

  await session.close();
  await client.close();
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: sdk/rust/examples/basic_chat.rs
//! Minimal end-to-end example: build a client, start a session, send one
//! prompt, and print streamed text deltas until the run completes.
//!
//! Requires `ANTHROPIC_API_KEY` in the environment. Run with:
//!
//! ```console
//! ANTHROPIC_API_KEY=sk-... cargo run -p rusty-harness-sdk --example basic_chat
//! ```

use std::sync::Arc;

use rusty_harness_sdk::protocol::{AgentEvent, AgentOutcome};
use rusty_harness_sdk::{Client, Session, SdkError};

#[tokio::main]
async fn main() -> Result<(), SdkError> {
    tracing_subscriber::fmt::init();

    let client = Client::builder()
        .register_integration(Arc::new(
            harness_integration_anthropic::AnthropicFactory,
        ))
        .build()
        .await?;

    let handle = client
        .session()
        .integration("anthropic", serde_json::json!({}))?
        .start()
        .await?;
    let session = Session::from(handle);

    let mut events = session.events();
    session.send("In one short sentence, what is a Rust trait?").await?;

    while let Some(event) = events.next().await {
        let envelope = event?;
        match envelope.event {
            AgentEvent::AssistantTextDelta { delta, .. } => print!("{delta}"),
            AgentEvent::Completed { outcome } => {
                println!();
                if !matches!(outcome, AgentOutcome::Success) {
                    eprintln!("run ended with outcome: {outcome:?}");
                }
                break;
            }
            AgentEvent::Failed { error } => {
                eprintln!("run failed: {error:?}");
                break;
            }
            _ => {}
        }
    }

    Ok(())
}
